lib/mastermind.rb

```diff
@@ -2,17 +2,20 @@
 require_relative './guess_validator'
 require_relative './guess_checker'
 require_relative './messages'
-require_relative './game_commands'
+require_relative './commands'
 
 class Mastermind
-  include GameCommands, Messages, UserInput
+  include Commands, Messages, UserInput
```

Comment:
Including the different modules like this makes it less clear where functions are defined / coming from.  I am actually in favor of just requiring, and maintaining the explicit calls with module name.   This also breaks the encapsulation we look for in modules, as when they are included the methods no longer live exclusively in the module, they are now part of the class. 

This becomes unmanageable at some point (I would argue that it is approaching that point now) due to the non-trivial amount of time it takes to understand what is going on in a relatively small codebase. 

A good example is `start_turn_message`.  In order to find where this code lives, I have to search through the files to find the method definition. 

Instead I would make each method in the Messages module a module level method and only require messages, not include them.  

`start_turn_message(@remaining_guesses)` becomes `Messages.start_turn_message(@remaining_guesses)` 

and 

```
def start_turn_message(remaining_guesses)
  puts "You have #{remaining_guesses} guesses remaining."
end

#becomes

def self.start_turn_message(remaining_guesses)
  puts "You have #{remaining_guesses} guesses remaining."
end
```

Doing it this way reduces coupling while increasing clarity.

Reply:
Awesome! Great to know - I wasn't sure how to use modules and was definitely concerned that it wouldn't be clear where these methods were coming from. Thanks!

Reply:
A module can be used anytime there is no state in the class/module.  Modules do not require you to new up an instantiation of a class. 
```ruby
class Cake
    def initialize(args = {})
        @flavor = args[:flavor] ||= :chocolate
    end

    def logFlavor
        puts "I am a #{@flavor} Cake!"
    end
end

module Pie
    def self.logFlavor
        puts "I am a apple pie!"
    end
end
```

```bash
irb(main):033:0> cake = Cake.new flavor: "carrot"
=> #<Cake:0x007fe789918390 @flavor="carrot">
irb(main):034:0> cake.logFlavor
I am a carrot Cake!
=> nil
irb(main):035:0> Pie.logFlavor
I am a apple pie!
=> nil
``